src/Language/Fixpoint/Solver/Interpreter.hs

```diff
@@ -0,0 +1,887 @@
+    bEnv              = filterBindEnv (\i _ _ -> memberIBindEnv i csBinds) (bs fi)
+    aEnv              = ae fi
+    γ                 = knowledge cfg fi  
+    s0                = EvalEnv sEnv mempty 
+
+---------------------------------------------------------------------------------------------- 
+-- | Step 1b: @mkCTrie@ builds the @Trie@ of constraints indexed by their environments 
+mkCTrie :: [(SubcId, SimpC a)] -> CTrie 
+mkCTrie ics  = mytracepp ("the trie is:   ") $ T.fromList [ (cBinds c, i) | (i, c) <- ics ]
+  where
+    cBinds   = L.sort . elemsIBindEnv . senv 
+
+---------------------------------------------------------------------------------------------- 
+-- | Step 2: @pleTrie@ walks over the @CTrie@ to actually do the incremental-PLE
+pleTrie :: CTrie -> InstEnv a -> IO InstRes   
+pleTrie t env = loopT env ctx0 diff0 Nothing res0 t 
+  where 
+    diff0        = []
+    res0         = M.empty 
+    ctx0         = initCtx env ((mkEq <$> es0) ++ (mkEq' <$> es0'))
+    es0          = L.filter (null . eqArgs) (aenvEqs   . ieAenv $ env)
+    es0'         = L.filter (null . smArgs) (aenvSimpl . ieAenv $ env)
+    mkEq  eq     = (EVar $ eqName eq, eqBody eq)
+    mkEq' rw     = (EApp (EVar $ smName rw) (EVar $ smDC rw), smBody rw)
+
+loopT :: InstEnv a -> ICtx -> Diff -> Maybe BindId -> InstRes -> CTrie -> IO InstRes
+loopT env ctx delta i res t = case t of 
+  T.Node []  -> return res
+  T.Node [b] -> loopB env ctx delta i res b
+  T.Node bs  -> (withAssms env ctx delta Nothing $ \ctx' -> do 
+                  (ctx'', res') <- ple1 env ctx' i res 
+                  foldM (loopB env ctx'' [] i) res' bs)
+
+loopB :: InstEnv a -> ICtx -> Diff -> Maybe BindId -> InstRes -> CBranch -> IO InstRes
+loopB env ctx delta iMb res b = case b of 
+  T.Bind i t -> loopT env ctx (i:delta) (Just i) res t
+  T.Val cid  -> withAssms env ctx delta (Just cid) $ \ctx' -> do 
+                  progressTick
+                  (snd <$> ple1 env ctx' iMb res) 
+
+
+withAssms :: InstEnv a -> ICtx -> Diff -> Maybe SubcId -> (ICtx -> IO b) -> IO b 
+withAssms env@(InstEnv {..}) ctx delta cidMb act = do 
+  let ctx'  = updCtx env ctx delta cidMb
+  act ctx' 
+
+-- | @ple1@ performs the PLE at a single "node" in the Trie 
+ple1 :: InstEnv a -> ICtx -> Maybe BindId -> InstRes -> IO (ICtx, InstRes)
+ple1 (InstEnv {..}) ctx i res = 
+  updCtxRes (mytracepp ("\nInstRes:  ") res) i <$> evalCandsLoop {-anfEnv-} M.empty ieCfg ctx ieKnowl ieEvEnv res
+{-    where
+      anfBEnv = filterBindEnv anfFilter ieBEnv 
+      anfFilter = (\i sym r -> (anfPrefix `isPrefixOfSym` sym) || (vvName `isPrefixOfSym` sym))
+
+      anfEnv = M.fromList ( Mb.mapMaybe unElabRefts (M.elems (beBinds anfBEnv)) )
+
+      unElabRefts (sym, sr) = let r = sr_reft sr in
+                                (,) (EVar sym) <$> (unElab <$> isSingletonE (reftBind r) (reftPred r))-}
+
+evalCandsLoop :: ConstMap -> Config -> ICtx -> Knowledge -> EvalEnv -> InstRes -> IO ICtx 
```

Comment:
Could we have documented here what the `res` parameter is for? This seems to be a difference with respect to the previous PLE.

Reply:
The InstRes object is being passed around, but it's no longer being used for anything in `evalsCandsLoop` or anything it calls. I'm removing it here. Thanks for pointing that out.